testsuites/syncgateway/functional/tests/test_xattrs_grants.py

```diff
@@ -573,15 +584,306 @@ def test_syncfunction_user_xattrs_dictionary_boolean(params_from_base_test_setup
+    mode = params_from_base_test_setup["mode"]
+    ssl_enabled = params_from_base_test_setup["ssl_enabled"]
+    sync_gateway_version = params_from_base_test_setup["sync_gateway_version"]
+    xattrs_enabled = params_from_base_test_setup["xattrs_enabled"]
+
+    if sync_gateway_version < "3.0.0":
+        pytest.skip('SGW version is not 3.0 and above')
+
+    if missing_type != "xattrs_disabled" and not xattrs_enabled:
+        pytest.skip('Test did not enable xattrs')
+    if missing_type == "xattrs_disabled" and xattrs_enabled:
+        pytest.skip('Param with xattrs disabled expected to disable xattrs')
+    sg_channel1_value = "abc"
+    sg_channels1 = [sg_channel1_value]
+    username = "autotest"
+    password = "password"
+    user_custom_channel1 = "channel1"
+    doc_xattrs_id = 'user_xattrs_format_0'
+    sg_conf_name = "custom_sync/sync_gateway_custom_sync"
+    sg_config = sync_gateway_config_path_for_mode(sg_conf_name, mode)
+
+    sg_client = MobileRestClient()
+
+    # 1. Set shared bucket access and import in config
+    # 2. Write a sync function to assign new channel using xattr value without adding xattrs key enabled
+    # 3. Verify SGW starts successfully
+    if missing_type == "user_xattrs_key":
+        temp_sg_config = replace_xattrs_sync_func_in_config(sg_config, user_custom_channel1, enable_xattrs_key=False)
+    else:
+        temp_sg_config = replace_xattrs_sync_func_in_config(sg_config, user_custom_channel1)
+
+    c_cluster = cluster.Cluster(config=cluster_config)
+    c_cluster.reset(sg_config_path=temp_sg_config)
+
+    cbs_ip = c_cluster.servers[0].host
+    cbs_bucket = c_cluster.servers[0].get_bucket_names()[0]
+    sdk_bucket = get_sdk_client_with_bucket(ssl_enabled, c_cluster, cbs_ip, cbs_bucket)
+    sg_client.create_user(sg_admin_url, sg_db, username, password, channels=sg_channels1)
+    auto_user = sg_client.create_session(url=sg_admin_url, db=sg_db, name=username)
+
+    # 4. create doc via SGW rest api
+    sg_docs = document.create_docs('user_xattrs_format', number=2)
+    sg_client.add_bulk_docs(url=sg_admin_url, db=sg_db, docs=sg_docs)
+
+    # 4. Update doc with SDK to add user xattr
+    if missing_type != "server_user_xattrs":
+        sdk_bucket.mutate_in(doc_xattrs_id, [SD.upsert(user_custom_channel1, sg_channel1_value, xattr=True, create_parents=True)])
+
+    # 6. Verify docs are not assigned to the channel which defined in user xattrs
+    sg_docs = sg_client.get_all_docs(url=sg_url, db=sg_db, auth=auto_user)["rows"]
+    sg_client.update_docs(url=sg_url, db=sg_db, docs=sg_docs, number_updates=1, auth=auto_user)
+    sg_docs = sg_client.get_all_docs(url=sg_url, db=sg_db, auth=auto_user)["rows"]
+    assert len(sg_docs) == 0, "docs assigned to the channel without user xattrs key"
+
+    raw_doc = sg_client.get_raw_doc(sg_admin_url, sg_db, doc_xattrs_id)
+
+    # 6. Perform raw GET only on one doc to ensure added xattr is imported
+    # raw_doc["_meta"]["xattrs"][user_custom_channel1]
```

Comment:
the numbers on the list kind of messed up, there are 2 #4 and 2 #6 ish. and also remove the line unneeded.

Reply:
done